cli: download files feature

ParthS007 opened the pull request:
closes #22

tiborsimko commented on cernopendata_client/cli.py:
```diff
@@ -100,25 +102,51 @@ def get_file_locations(server, recid, doi, title, protocol, expand):
+@click.option("--recid", type=click.INT, help="Record ID")
+@click.option("--doi", help="Digital Object Identifier.")
+@click.option("--title", help="Record title")
+@click.option(
```
Since we don't have `root` supported yet, maybe don't introduce this option... or, better, switch the defaults to use `http` and if people select `root` then reply that this is not yet supported and do `sys.exit(1)`.

tiborsimko replied:
(we could also with default to 'http' over 'root' for the other `get-file-locations` function)

tiborsimko commented on cernopendata_client/cli.py:
```diff
@@ -100,25 +102,51 @@ def get_file_locations(server, recid, doi, title, protocol, expand):
+    for file_location in file_locations:
+        file_dest = path + "/" + file_location.split("/")[-1]
+        with open(file_dest, "wb") as f:
+            print("File Downloading")
```
It would be nice to introduce some output formatting cosmetics, e.g. currently people see:

```console
$ cernopendata-client download-files --recid 5500 --protocol http
Creation of the directory 5500 failed
File Downloading
File Downloading kiB (100%)
File Downloading81 kiB (100%)
File Downloading kiB (100%)
File Downloading14 kiB (100%)
File Downloading15 kiB (100%)
File Downloading kiB (100%)
File Downloading kiB (100%)
File Downloading kiB (100%)
File Downloading kiB (100%)
File Downloading17 kiB (100%)
Downloading: 90/90 kiB (100%)
```

I would expect to see the URI or at least the file name that is being currently downloaded.  Also, the total count, so that people can estimate the ETA.  Somethhing like this:

```
==> Downloading file 1 of 10:  demoanalyzer_cfg_level3data.py
Downloading 123/123456789 KiB (1.23%)
==> Downloading file 2 of 10: demoanalyzer_cfg_level3MC.py
...
```

tiborsimko commented on setup.py:
```diff
@@ -45,7 +45,7 @@
 ]
 
-install_requires = ["click>=7,<8", "requests==2.22.0"]
+install_requires = ["click>=7,<8", "requests==2.22.0", "pycurl>=7"]
```
Travis CI fails with:

```
src/pycurl.h:202:30: fatal error: gnutls/gnutls.h: No such file or directory
```

You may need to amend `.travis.yml` to preinstall `gnutls` libraries.

tiborsimko commented on cernopendata_client/helper.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
`helper.py` is a rather generic name... since it can help with anything.  Better invent some more targeted name. 

- One option is to put it in `download.py` and have all the functions related to downloads sitting in that file.
- Another option is to put it in `tui.py` and have all the "text user interface" functions sitting in that file.

The former is probably better, since we wouldn't be easily able to reuse this function elsewhere in other contexts than downloading.

WDYT?

ParthS007 replied:
The reason why I choose `helper.py` is that I thought we will be adding other functions that are on their own and don't require any data apart from arguments and we can add other functions not just limited to download functionality in that file which are doing some calculations or providing customized output on the terminal.

I think `tui.py` will fulfill this purpose and more understandable than `helper.py` or `download.py`. 

going with `tui.py` then 👍

tiborsimko replied:
I meant that if this function is not going to be called by anybody except the `download-...` set of commands, then `download.py` may be appropriate location. If it would be called by `download-...` and `get-...` etc commands, for example styling general output messages, then `tui.py` may be more appropriate location. I'm OK with both, as `cernopendata-client` will grow, we may actually be revamping things later. (E.g. `reana-dev` started as one single file, but as the number of commands grew, we separate each command set into its own file.)

ParthS007 replied:
I agree and it sounds perfect. 👍

tiborsimko commented on cernopendata_client/cli.py:
```diff
@@ -81,7 +83,7 @@ def get_record(server, recid, doi, title, output_fields):
     "--protocol",
-    default="root",
+    default="http",
     type=click.Choice(["root", "http"]),
```
Please order alphabetically, `type=click.Choice(["http", "root"]),` here and in the other function.

tiborsimko commented on cernopendata_client/cli.py:
```diff
@@ -100,25 +102,67 @@ def get_file_locations(server, recid, doi, title, protocol, expand):
+
+    path = str(recid)
+    try:
+        os.mkdir(path)
```
Please check first whether the directory does not already exists; there is no need to create it then.  Users would see a message like `Creation of the directory 5500 failed` and might be confused.

tiborsimko reviewed (commented):
Works well :+1: I left two more cosmetic comments. 

@ParthS007 Shall we merge before adding tests? Or after?

ParthS007 commented:
@tiborsimko  I can write the test for testing the download helper in `tui.py` now. Please provide me the location of the file we discussed in person.

tiborsimko commented:
> I can write the test for testing the download helper in `tui.py` now. Please provide me the location of the file we discussed in person.

Good then! The tests for `tui.py` should go to `tests/test_tui.py` file. You can emulate `test_validator.py` that contain tests for `validator.py`.

tiborsimko commented:
> > I can write the test for testing the download helper in `tui.py` now. Please provide me the location of the file we discussed in person.
> 
> Good then! The tests for `tui.py` should go to `tests/test_tui.py` file. You can emulate `test_validator.py` that contain tests for `validator.py`.

We were testing `http://opendata.cern.ch/record/1050/files/2010RunBlumi.txt` but record 1050 has three files; let's perhaps use another file which has only one?  Example: `http://opendata.cern.ch/record/3005/files/0d0714743f0204ed3c0144941e6ce248.configFile.py`

ParthS007 commented:
Testing one file seems fine to me for now, once I implement a low-level download_function, then we can add other test cases as well. what do you think?

ParthS007 commented:
@tiborsimko I added basic tests for testing CLI commands and since we are not calling directly `show_download_progress`, I thought of writing a test for `download-files` command instead of `show_download_progress`.

Is this the right way to approach this?

Please let me know, I will do the changes if required.

For `get_record` and `get_file_locations`, I added basic tests for checking the output type on the basis of `recid`.

tiborsimko commented on tests/test_cli.py:
```diff
@@ -0,0 +1,34 @@
+# cernopendata-client is free software; you can redistribute it and/or modify
+# it under the terms of the GPLv3 license; see LICENSE file for more details.
+
+"""cernopendata-client text user interface(tui) test."""
```
This test suite will grow, so please create three test files, one for each CLI command:

- `tests/test_get_record.py`
- `tests/test_get_file_locations.py`
- `tests/test_download_files.py`

?

ParthS007 replied:
Creating three test files sounds good, I was thinking of one more folder like

- `tests/tests_cli/test_get_record.py`
- `tests/tests_cli/test_get_file_locations.py`
- `tests/tests_cli/test_download_files.py`

which one looks fine?

tiborsimko replied:
Yes, that's also OK, especially if we are going to split `cli.py` into `cli/command1.py`, `cli/command2.py` one day...

tiborsimko commented on tests/test_cli.py:
```diff
@@ -0,0 +1,34 @@
+    """Test get_record() command."""
+    test_get_record = CliRunner()
+    test_result = test_get_record.invoke(get_record, ["--recid", 3005])
+    assert type(test_result.output) == str
```
Instead of just asserting type string, you could also assert exit code (should be 0) and, more importantly, the presence of some expected text in the output.

Example:

```python
    assert '"title": "Configuration file for LHE step HIG-Summer11pLHE-00114_1_cfg.py"' in test_result.output
    assert test_result.exit_code == 0
```

The same for the other commands, e.g. when listing file locations, assert presence of a known one what the command should return, etc.

tiborsimko reviewed (commented):
> Is this the right way to approach this?

Yes, I like it :+1: I left two comments about (i) splitting tests for each command into its own file, and about (ii) enriching asserts by looking for expected strings.

tiborsimko commented on tests/tests_cli/test_get_file_locations.py:
```diff
@@ -0,0 +1,22 @@
+    test_result = test_get_file_locations.invoke(get_file_locations, ["--recid", 3005])
+    assert test_result.exit_code == 0
+    assert type(test_result.output) == str
+    assert 'configFile.py' in test_result.output
```
Better to test the full name `0d0714743f0204ed3c0144941e6ce248.configFile.py` so that we are sure we did not get a file from another record.  (There are many `xxxxx.configFile.py` ones.)

tiborsimko commented on tests/tests_cli/test_download_files.py:
```diff
@@ -0,0 +1,21 @@
+    test_download_files = CliRunner()
+    test_result = test_download_files.invoke(download_files, ["--recid", 3005])
+    assert test_result.exit_code == 0
+    assert test_result.output.endswith("\nDownload completed!\n")
```
You can also test whether `3005/0d0714743f0204ed3c0144941e6ce248.configFile.py` file is well present and whether its size is 3644 bytes.

BTW it may be necessary before running this test case to delete that file if it is present... so that we don't get false positives...

tiborsimko commented on tests/tests_cli/test_download_files.py:
```diff
@@ -0,0 +1,27 @@
+    assert test_result.exit_code == 0
+    assert os.path.isfile(test_file) is True
+    assert os.path.getsize(test_file) == 3644
+    assert test_result.output.endswith("\nDownload completed!\n")
```
Nice, please also remove the file after test is done as the last step.

BTW I get `black' warnings on the two test files...

tiborsimko replied:
Note also the first comment, please add:

```
    if os.path.isfile(test_file):
        os.remove(test_file)
```

as the last instructions in the `test_download_files()` function, so that we end up with a blank slate again.

After this is done, we are ready to go :tada:
